### ISSUE TITLE
[Core] Allow NaN target values at `fit`/`preprocess`, dropping them only before model fitting

### ISSUE THREAD

**Issue by simonez-tuidi** [Nixtla/mlforecast#704]
### Description
Today `mlforecast` hard-rejects any null/NaN in the target column. The single gate is in
`TimeSeries._fit`:

```python
# mlforecast/core.py:551-554
validate_format(df, id_col, time_col, target_col)
validate_freq(df[time_col], self.freq)
if ufp.is_nan_or_none(df[target_col]).any():
    raise ValueError(f"{target_col} column contains null values.")
```

Because every entry point routes through `TimeSeries._fit`, this blocks NaN targets everywhere:
`MLForecast.preprocess` → `TimeSeries.fit_transform` (`mlforecast/core.py:1216`), `MLForecast.fit`
(→ `preprocess`, `mlforecast/forecast.py:1211`), `MLForecast.history_warmup`,
`MLForecast.cross_validation`, `LightGBMCV.fit` (`mlforecast/lgb_cv.py:246`) and the distributed
`DistributedMLForecast` variants (`mlforecast/distributed/forecast.py:227`).

The behaviour is locked in by `tests/test_core.py:361-367`:

```python
series_with_nulls = series.copy()
series_with_nulls.loc[1, "y"] = np.nan
with pytest.raises(Exception) as exec:
    ts.fit_transform(series_with_nulls, id_col="unique_id", time_col="ds", target_col="y")
assert "y column contains null values" in str(exec.value)
```

(also snapshotted in `docs/core.html.md:250`).

**Request:** allow NaN in the target at `fit`/`preprocess`/`history_warmup`. NaN rows should still
participate in feature construction *positionally* (so the time grid stays intact), be *excluded*
from the statistics computed over them (they must not inflate the denominator of a rolling mean,
count, std, …), and be dropped from `X`/`y` immediately before the estimator is fitted.

Ideally opt-in, e.g. `allow_null_target: bool = False` on `preprocess`/`fit`/`history_warmup`, so
current behaviour is unchanged by default.

## Use case

We forecast retail/product-level demand. A product typically has a **life-cycle**: it does not exist,
or is out of range/delisted, for part of the history. The intervals where it does not exist and the
intervals where it exists but sold nothing are two completely different signals.

`mlforecast` requires a gap-free, regular time grid — `validate_df` raises on gaps:

```python
# mlforecast/data_validation.py:221-228
has_issues, bad = validate_continuity(df, id_col, time_col, freq)
if has_issues:
    raise ValueError(
        f"Series contain missing or duplicate timestamps with the specified freq {freq}\n"
        f"Affected series: {bad_ids}\n"
        f"Consider using the fill_gaps parameter or preprocessing your data."
    )
```

(and `validate_update_df` at `mlforecast/data_validation.py:192-199` for `update`; pooled
`global_`/`groupby`/`partition_by` transforms additionally *require* a dense grid — see the
`validate_data=False` warning at `mlforecast/forecast.py:263-271`.)

So the two requirements combine into a forced-imputation trap: we must fill the gaps, and the filler
cannot be NaN, so it becomes `0`. That `0` is then indistinguishable from a genuine zero-sales week.
Concretely:

- A rolling mean over a 12-week window that includes 8 weeks of "product did not exist" reads as a
  mean over 12 observations mostly at zero, instead of a mean over the 4 real observations.
- Counts / `min_samples` gates see 12 samples where there are 4, so features look "mature" when they
  are not.
- Every model trained on this learns "post-launch demand looks like a ramp from zero", which is an
  artifact of the imputation.
- Rows where the product does not exist are also not legitimate training rows — the target itself is
  not defined there — but there is no way to say so, so we either train on fabricated zeros or filter
  the preprocessed frame ourselves with a hand-maintained mask, which is easy to get wrong and has to
  be re-derived for every horizon in the `max_horizon` case.

The workaround we would want to write — put NaN where the product does not exist, let the transforms
skip it, drop those rows before fitting — is exactly what NaN semantics give for free, and it is what
we would get from SQL window functions over the same data.

## Why this looks close to free

Most of the machinery already exists and already has the desired semantics.

**1. "Drop NaN target rows before fitting" is already implemented.** `TimeSeries._transform` drops
target-null rows unconditionally, *even with* `dropna=False`:

```python
# mlforecast/core.py:936-949
target_nulls = np.isnan(target)
if target_nulls.ndim == 2:
    # target nulls for each horizon are dropped in MLForecast.fit_models
    # we just drop rows here for which all the target values are null
    target_nulls = target_nulls.all(axis=1)
if dropna:
    ...
    keep_rows = ~(feature_nulls | target_nulls)
else:
    # we always want to drop rows with nulls in the target
    keep_rows = ~target_nulls
```

**2. Per-horizon NaN masking is already implemented** for the multi-output path, in both
`TimeSeries._horizon_train_data` (`mlforecast/core.py:1175`) and `MLForecast.fit_models`
(`mlforecast/forecast.py:864`):

```python
valid = ~np.isnan(y_h)
...
X_h = ufp.filter_with_mask(X_h, valid)
y_h = y_h[valid]
```

so "features computed on the full grid, model fitted on the valid subset" is an established pattern
in this codebase, not a new one.

**3. The pooled transforms are already NaN-skipping, by design and by documentation.**
`_RollingBase.__init__` documents `min_samples` in exactly the terms this request needs
(`mlforecast/lag_transforms.py:506-509`, and `1026-1029` for the seasonal variant):

> In pooled mode (`global_=True`, `groupby` or `partition_by`), `min_samples` counts total **non-NaN
> observations** across **all series** in the bucket within the rolling window, with no capping.

and the implementations agree — the generic slow path drops NaN before the statistic
(`mlforecast/lag_transforms.py:598-608`):

```python
vals = y_b[mask]
vals = vals[~np.isnan(vals)]
if len(vals) >= min_samples and len(vals) > 0:
    feat_u[k] = self._window_stat(vals)
```

the aggregate fast paths carry a non-NaN `counts` array and divide by it
(`_rolling_mean_from_agg`, `mlforecast/lag_transforms.py:616-633`), and `PooledState` builds its
per-timestamp aggregates from
a validity mask (`mlforecast/pooled.py:195`, `mlforecast/pooled.py:243`, `mlforecast/pooled.py:286`,
plus the incremental update at `mlforecast/pooled.py:959-1004`).

This is verifiable today by patching out the guard at `core.py:553`. Two series `a`/`b`, daily,
`RollingMean(window_size=3, min_samples=1, global_=True)`, with `a` NaN on 01-09..01-11:

```
 unique_id         ds     y  lag1  global_rolling_mean_lag1_window_size3_min_samples1
         a 2020-01-08   8.0   7.0                                                55.5
         a 2020-01-12  12.0   NaN                                               109.0   <- mean of b's 108,109,110
         a 2020-01-13  13.0  12.0                                                85.5
```

The NaN rows are dropped from the output automatically (point 1), and the window at 01-12 averages
only the three values that actually exist rather than dividing by six. That is precisely the
requested behaviour, already working.

## The part that does need work

The **local (per-series) path** delegates to `coreforecast`, which propagates NaN *by default*, so a
single NaN poisons the rest of the series:

```python
y = np.array([1., 2., np.nan, 4., 5., 6.])
ct.RollingMean(lag=1, window_size=3, min_samples=1).transform(ga)
# -> [nan  1.  1.5  nan  nan  nan]      expected [nan 1. 1.5 1.5 3. 4.5]
ct.ExpandingMean(lag=1).transform(ga)
# -> [nan  1.  1.5  nan  nan  nan]
ct.Lag(lag=1).transform(ga)
# -> [nan  1.  2.  nan  4.  5.]         correct: the lagged value genuinely is NaN
```

End to end (single series, `lags=[1]`, `RollingMean(window_size=3, min_samples=1)`, NaN on rows
8-10, guard patched out): the rolling mean is `NaN` for **every** row from row 11 to the end, while
the same configuration with `global_=True` is correct.

**But this is opt-out, not a missing capability.** `coreforecast` 0.0.18 already takes a `skipna`
argument on every one of these transforms, and it produces exactly the right answer — `mlforecast`
simply never sets it:

```python
ct.RollingMean(1, 3, 1, skipna=True).transform(ga)   # -> [nan 1.  1.5 1.5 3.  4.5]   matches pandas
ct.ExpandingMean(1, skipna=True).transform(ga)       # -> [nan 1.  1.5 1.5 2.33 3.  ]
ct.ExponentiallyWeightedMean(1, .5, skipna=True).transform(ga)  # -> [nan 1. 1.5 1.5 2.75 3.875]
ct.SeasonalRollingMean(1, 2, 2, 1, skipna=True).transform(ga)   # -> [nan 1. 2.  1.  3.  5.  ]
```

And plumbing it through looks nearly mechanical: `_BaseLagTransform._set_core_tfm`
(`mlforecast/lag_transforms.py:133-143`) already forwards every non-pooled `__init__` argument to the
`coreforecast` class by name, so adding `skipna` to `_RollingBase.__init__` /
`_Seasonal_RollingBase.__init__` / `_ExpandingBase.__init__` / `ExponentiallyWeightedMean.__init__`
would reach `coreforecast` with no change to `_set_core_tfm` itself. `_get_name`
(`mlforecast/lag_transforms.py:145-167`) only appends parameters whose value differs from the
default, so `skipna=False` keeps every existing feature name byte-identical.

Two caveats found while checking this:

- **`coreforecast`'s `update()` ignores `skipna`** for `ExpandingMean`, `ExpandingStd`,
  `ExpandingMin`, `ExpandingMax` and `ExponentiallyWeightedMean` — exactly the five whose `update()`
  is implemented in Python rather than delegating to `_lib`. `transform()` is correct for all of
  them, so training features would be right and recursive prediction would silently diverge. The
  rolling/seasonal/quantile transforms are fine. Filed upstream against `coreforecast`; `mlforecast`
  should either wait for that fix or restrict `skipna` to the transforms whose `update()` honors it.
- **Version floor.** `skipna` does not exist in `coreforecast` 0.0.16 (zero occurrences in
  `rolling.py`, `expanding.py`, `lag_transforms.py`, `scalers.py`); it is present throughout 0.0.18.
  `pyproject.toml:27` currently declares `coreforecast>=0.0.15`, which would need raising.

Other things that would need a decision:

- **`Lag` itself** — `lag1` of a NaN is legitimately NaN; those rows get dropped by `dropna=True`
  as usual. Worth stating explicitly in the docs.
- **Whether `skipna` is per-transform or a single dataset-level switch.** Per-transform matches
  `coreforecast` and lets a user keep NaN-propagating semantics where they want them; a single
  `preprocess(..., skipna=True)` is friendlier but has to fan out to every transform anyway.
- **Target transforms** (`Differences`, `AutoDifferences`, `LocalStandardScaler`, … in
  `mlforecast/target_transforms.py`) — the `coreforecast` scalers behind these already accept
  `skipna` (`LocalMinMaxScaler`, `LocalStandardScaler`, `LocalRobustScaler`; `LocalStandardScaler(skipna=True)`
  on `[1, 2, nan, 4, 5]` correctly yields `mean=3.0, std=1.581`, vs `nan/nan` with the default), so
  the same threading applies. `LocalBoxCoxScaler` and `Difference` do not take `skipna`.
- **Recursive `predict`/`update`** — history that contains NaN feeding the update rules; the pooled
  incremental update already handles it (`mlforecast/pooled.py:959-1004`), the local
  `_core_tfm.update` path does not.
- **Feature nulls are already tolerated with a warning**, not a raise
  (`mlforecast/core.py:1576`: `warnings.warn(f"Found null values in {', '.join(cols_with_nulls)}.")`),
  which suggests the target could reasonably follow the same permissive-with-a-warning philosophy.

## Proposed shape

1. Replace the unconditional raise at `mlforecast/core.py:553-554` with a check gated on a new
   `allow_null_target=False` argument threaded through `TimeSeries.fit_transform` /
   `TimeSeries.history_warmup` and exposed on `MLForecast.preprocess` / `fit` /
   `history_warmup` / `cross_validation` (and the distributed + `LightGBMCV` equivalents).
2. When enabled: keep NaN rows on the grid for feature computation, rely on the existing
   `keep_rows = ~target_nulls` drop (`mlforecast/core.py:948-949`) and per-horizon `valid` masks
   (`mlforecast/core.py:1175`, `mlforecast/forecast.py:864`) to keep them out of `fit`.
3. Add `skipna` to the local transforms in `mlforecast/lag_transforms.py` and let the existing
   `_set_core_tfm` forwarding carry it into `coreforecast`, raising a targeted error for the
   transforms whose `coreforecast` `update()` does not yet honor it — not a blanket rejection of the
   whole dataframe.
4. Document that `forecast_fitted_values()` will not contain the NaN-target rows, same as the
   existing `dropna` behaviour.

Happy to work on a PR if the maintainers agree on the shape — in particular on whether `skipna`
should be per-transform (mirroring `coreforecast`) or a single switch on `preprocess`/`fit`, and
whether to ship before or after the upstream `coreforecast` `update()` fix.
